File: frontend/src/schemas/trip.ts
/**
 * Trip schemas
 * 
 * Schemas for trip-related entities:
 * - TripTemplate, TripOccurrence, RecommendedTripOccurrence
 * - Response wrapper schemas for trip endpoints
 * - Special response schemas (HealthCheckResponseSchema, SchemaInfoResponseSchema)
 * 
 * Note: Uses z.lazy() to handle circular dependencies between TripTemplate and TripOccurrence.
 */

import { z } from 'zod';
import { TripStatusSchema, DifficultyLevelSchema } from './base';
import { GuideSchema, CompanySchema, TripTypeSchema, CountrySchema, TagSchema } from './resources';

// ============================================
// CIRCULAR REFERENCE SCHEMAS (using z.lazy())
// ============================================

/**
 * TripTemplate schema (defined first, references TripOccurrenceSchema via lazy)
 * Note: Circular reference fields (occurrences) use z.lazy() to prevent circular dependency issues
 */
export const TripTemplateSchema: z.ZodType<any> = z.object({
  id: z.number(),
  title: z.string(),
  titleHe: z.string(),
  description: z.string(),
  descriptionHe: z.string(),
  shortDescription: z.string().nullable().optional(),
  shortDescriptionHe: z.string().nullable().optional(),
  imageUrl: z.string().nullable().optional(),
  basePrice: z.union([z.number(), z.string()]).transform((val) => typeof val === 'string' ? parseFloat(val) : val),
  singleSupplementPrice: z.union([z.number(), z.string()]).transform((val) => typeof val === 'string' ? parseFloat(val) : val).nullable().optional(),
  typicalDurationDays: z.number(),
  defaultMaxCapacity: z.number(),
  difficultyLevel: DifficultyLevelSchema,
  companyId: z.number(),
  tripTypeId: z.number().nullable().optional(),
  primaryCountryId: z.number().nullable().optional(),
  isActive: z.boolean(),
  properties: z.record(z.string(), z.any()).nullable().optional(),
  createdAt: z.string(),
  updatedAt: z.string(),
  // Optional relations (occurrences uses lazy for circular reference, others are validated)
  company: CompanySchema.nullable().optional(),
  tripType: TripTypeSchema.nullable().optional(),
  primaryCountry: CountrySchema.nullable().optional(),
  countries: z.array(CountrySchema).optional(),
  tags: z.array(TagSchema).optional(),
  // Lazy reference to TripOccurrenceSchema (defined below)
  occurrences: z.lazy(() => z.array(TripOccurrenceSchema)).optional(),
}).passthrough(); // Allow additional fields from backend

/**
 * TripOccurrence schema
 * Note: Circular reference fields (template) use z.lazy() to prevent circular dependency issues
 */
export const TripOccurrenceSchema: z.ZodType<any> = z.object({
  id: z.number(),
  tripTemplateId: z.number(),
  guideId: z.number().nullable().optional(),
  startDate: z.string(),
  endDate: z.string(),
  priceOverride: z.union([z.number(), z.string()]).transform((val) => typeof val === 'string' ? parseFloat(val) : val).nullable().optional(),
  singleSupplementOverride: z.union([z.number(), z.string()]).transform((val) => typeof val === 'string' ? parseFloat(val) : val).nullable().optional(),
  maxCapacityOverride: z.number().nullable().optional(),
  spotsLeft: z.number(),
  status: TripStatusSchema,
  imageUrlOverride: z.string().nullable().optional(),
  notes: z.string().nullable().optional(),
  notesHe: z.string().nullable().optional(),
  properties: z.record(z.string(), z.any()).nullable().optional(),
  createdAt: z.string(),
  updatedAt: z.string(),
  // Computed fields (backend sends Decimal as string)
  effectivePrice: z.union([z.number(), z.string()]).transform((val) => typeof val === 'string' ? parseFloat(val) : val).nullable().optional(),
  effectiveMaxCapacity: z.number().nullable().optional(),
  effectiveImageUrl: z.string().nullable().optional(),
  durationDays: z.number().nullable().optional(),
  // Optional relations (template uses lazy for circular reference, guide is validated)
  template: z.lazy(() => TripTemplateSchema).nullable().optional(),
  guide: GuideSchema.nullable().optional(),
}).passthrough(); // Allow additional fields from backend

/**
 * RecommendedTripOccurrence schema
 * Note: Includes all TripOccurrence fields plus recommendation-specific fields
 * Uses z.lazy() for template field to handle circular reference
 * Uses .passthrough() to allow additional fields from backend
 */
export const RecommendedTripOccurrenceSchema: z.ZodType<any> = z.object({
  id: z.number(),
  tripTemplateId: z.number(),
  guideId: z.number().nullable().optional(),
  startDate: z.string(),
  endDate: z.string(),
  priceOverride: z.union([z.number(), z.string()]).transform((val) => typeof val === 'string' ? parseFloat(val) : val).nullable().optional(),
  singleSupplementOverride: z.union([z.number(), z.string()]).transform((val) => typeof val === 'string' ? parseFloat(val) : val).nullable().optional(),
  maxCapacityOverride: z.number().nullable().optional(),
  spotsLeft: z.number(),
  status: TripStatusSchema,
  imageUrlOverride: z.string().nullable().optional(),
  notes: z.string().nullable().optional(),
  notesHe: z.string().nullable().optional(),
  properties: z.record(z.string(), z.any()).nullable().optional(),
  createdAt: z.string(),
  updatedAt: z.string(),
  // Computed fields (backend sends Decimal as string)
  effectivePrice: z.union([z.number(), z.string()]).transform((val) => typeof val === 'string' ? parseFloat(val) : val).nullable().optional(),
  effectiveMaxCapacity: z.number().nullable().optional(),
  effectiveImageUrl: z.string().nullable().optional(),
  durationDays: z.number().nullable().optional(),
  // Optional relations (template uses lazy for circular reference, guide is validated)
  template: z.lazy(() => TripTemplateSchema).nullable().optional(),
  guide: GuideSchema.nullable().optional(),
  // Recommendation-specific fields
  match_score: z.number(),
  match_details: z.array(z.string()).optional(), // Sometimes might be missing
  is_relaxed: z.boolean().optional(),
}).passthrough(); // Allow additional fields from backend

// ============================================
// RESPONSE WRAPPER SCHEMAS
// ============================================

/**
 * Generic ApiResponse schema factory
 */
function createApiResponseSchema<T extends z.ZodTypeAny>(dataSchema: T) {
  return z.object({
    success: z.boolean(),
    data: dataSchema.optional(),
    count: z.number().nullable().optional(),
    total: z.number().nullable().optional(),
    error: z.string().nullable().optional(), // Can be null or undefined
    message: z.string().nullable().optional(), // Can be null or undefined
    api_version: z.string().nullable().optional(),
  }).passthrough(); // Allow additional fields like primary_count, relaxed_count, etc.
}

/**
 * Trip response schemas
 */
export const TripTemplateArrayResponseSchema = createApiResponseSchema(z.array(TripTemplateSchema));
export const TripTemplateResponseSchema = createApiResponseSchema(TripTemplateSchema);
export const TripOccurrenceArrayResponseSchema = createApiResponseSchema(z.array(TripOccurrenceSchema));
export const TripOccurrenceResponseSchema = createApiResponseSchema(TripOccurrenceSchema);
export const RecommendedTripOccurrenceArrayResponseSchema = createApiResponseSchema(z.array(RecommendedTripOccurrenceSchema));

// ============================================
// SPECIAL RESPONSE SCHEMAS
// ============================================

/**
 * Health check response schema (non-standard format)
 * Returns: { status, service, version, schema, database }
 */
export const HealthCheckResponseSchema = z.object({
  status: z.string(),
  service: z.string(),
  version: z.string(),
  schema: z.string(),
  database: z.object({
    trip_occurrences: z.number(),
    trip_templates: z.number(),
    countries: z.number(),
    guides: z.number(),
    tags: z.number(),
    trip_types: z.number(),
  }),
});

/**
 * Schema info response schema
 */
export const SchemaInfoResponseSchema = z.object({
  success: z.boolean(),
  schema_version: z.string(),
  statistics: z.object({
    companies: z.number(),
    templates: z.number(),
    occurrences: z.number(),
    active_occurrences: z.number(),
  }),
  endpoints: z.record(z.string(), z.string()),
  error: z.string().optional(),
});

File: backend/app/schemas/trip.py
"""
Schemas for Trip Models (TripTemplate, TripOccurrence, Trip)
"""

from datetime import datetime, date
from typing import Optional, List
from decimal import Decimal
from pydantic import ConfigDict
from app.schemas.base import BaseSchema
from app.schemas.resources import CompanySchema, TripTypeSchema, CountrySchema, TagSchema, GuideSchema


class TripTemplateSchema(BaseSchema):
    """
    Schema for TripTemplate API responses.
    
    All fields use snake_case internally but serialize to camelCase.
    Example: image_url → imageUrl in JSON response
    """
    id: int
    title: str
    title_he: str  # Automatically becomes "titleHe" in JSON
    description: str
    description_he: str  # Automatically becomes "descriptionHe" in JSON
    short_description: Optional[str] = None  # Automatically becomes "shortDescription"
    short_description_he: Optional[str] = None  # Automatically becomes "shortDescriptionHe"
    image_url: Optional[str] = None  # Automatically becomes "imageUrl"
    base_price: Decimal  # Automatically becomes "basePrice"
    single_supplement_price: Optional[Decimal] = None  # Automatically becomes "singleSupplementPrice"
    typical_duration_days: int  # Automatically becomes "typicalDurationDays"
    default_max_capacity: int  # Automatically becomes "defaultMaxCapacity"
    difficulty_level: int  # Automatically becomes "difficultyLevel"
    company_id: int  # Automatically becomes "companyId"
    trip_type_id: Optional[int] = None  # Automatically becomes "tripTypeId"
    primary_country_id: Optional[int] = None  # Automatically becomes "primaryCountryId"
    is_active: bool  # Automatically becomes "isActive"
    properties: Optional[dict] = None  # JSONB - no transformation needed
    created_at: datetime  # Automatically becomes "createdAt"
    updated_at: datetime  # Automatically becomes "updatedAt"
    
    # Relations (optional, included when needed)
    company: Optional[CompanySchema] = None
    trip_type: Optional[TripTypeSchema] = None  # Automatically becomes "tripType"
    primary_country: Optional[CountrySchema] = None  # Automatically becomes "primaryCountry"
    # Note: countries and tags are excluded by default to avoid lazy loading issues
    # They can be included explicitly if needed via model_validate with proper eager loading
    countries: List[CountrySchema] = []
    tags: List[TagSchema] = []
    
    model_config = ConfigDict(from_attributes=True, populate_by_name=True)


class TripOccurrenceSchema(BaseSchema):
    """Schema for TripOccurrence API responses"""
    id: int
    trip_template_id: int
    guide_id: Optional[int] = None
    start_date: date
    end_date: date
    price_override: Optional[Decimal] = None
    single_supplement_override: Optional[Decimal] = None
    max_capacity_override: Optional[int] = None
    spots_left: int
    status: str
    image_url_override: Optional[str] = None
    notes: Optional[str] = None
    notes_he: Optional[str] = None
    properties: Optional[dict] = None
    created_at: datetime
    updated_at: datetime
    
    # Computed fields (from hybrid properties)
    effective_price: Optional[Decimal] = None
    effective_max_capacity: Optional[int] = None
    effective_image_url: Optional[str] = None
    duration_days: Optional[int] = None
    
    # Relations (optional, included when needed)
    template: Optional[TripTemplateSchema] = None
    guide: Optional[GuideSchema] = None
    
    model_config = ConfigDict(from_attributes=True)


